File: day10.hs
import qualified Data.List as List
import qualified Data.Text as T
import qualified Data.Text.IO as TIO

parse :: T.Text -> [Int]
parse content = map (read . T.unpack) $ T.lines $ content

diffs :: [Int] -> [Int]
diffs vals = map (\(a, b) -> b - a) $ zip (0:sortedContent) sortedContent
    where
        sortedContent = List.sort vals

answer :: [Int] -> Int
answer content = diff1 * diff3
    where
        ds = diffs content
        diff1 = length $ filter ((==) 1) ds
        diff3 = 1 + (length $ filter ((==) 3) ds)

main :: IO ()
main = do
    content <- TIO.readFile "day10.txt"
    let input = parse content
    putStrLn $ show $ answer input

File: python/day10.py
#!/usr/bin/env python3

with open("../input/day10.txt") as f:
    data = f.read()
data = list(map(int, data.splitlines()))

l = sorted(data)
l[0:0] = [0]
l.append(l[-1] + 3)
diffs = [l[i] - l[i - 1] for i in range(1, len(l))]
#print(repr(diffs))
n1 = len([x for x in diffs if x == 1])
n3 = len([x for x in diffs if x == 3])
#print(n1, n3)
print(n1 * n3)
